### PR TITLE
bpo-41825: restructure docs for the os.wait*() family

### LINKED ISSUE #85991: os.waitid() documentation needs TLC
BPO | [41825](https://bugs.python.org/issue41825)
--- | :---
Nosy | @birkenfeld, @iritkatriel
PRs | <li>python/cpython#22356</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-09-21.16:41:50.762>
labels = ['easy', '3.9', '3.10', '3.11', 'type-feature', 'library', 'docs']
title = 'os.waitid() documentation needs TLC'
updated_at = <Date 2021-06-24.23:44:30.315>
user = 'https://github.com/birkenfeld'
```

bugs.python.org fields:
```python
activity = <Date 2021-06-24.23:44:30.315>
actor = 'iritkatriel'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'Library (Lib)']
creation = <Date 2020-09-21.16:41:50.762>
creator = 'georg.brandl'
dependencies = []
files = []
hgrepos = []
issue_num = 41825
keywords = ['patch', 'easy']
message_count = 2.0
messages = ['377263', '396516']
nosy_count = 3.0
nosy_names = ['georg.brandl', 'docs@python', 'iritkatriel']
pr_nums = ['22356']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue41825'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### PR DESCRIPTION
**Pull request by birkenfeld**
Mostly fixes up the specific docs of `waitid`, so that basic usage and all flags are explained without having to consult the manpage.

While doing that, I noticed a few issues with the other `wait*` functions.  I tried to reorder them and `W*` constants to be sequential, and clearly mark which flag can be used for which function.

https://bugs.python.org/issue41825

Fixes #85991

### CODE REVIEW

File: Doc/library/os.rst [hunk 1]
```diff
@@ -4135,8 +4096,11 @@ written in Python, such as a mail server's external command delivery program.
    ``-1``, status is requested for any process in the process group ``-pid`` (the
    absolute value of *pid*).
 
-   An :exc:`OSError` is raised with the value of errno when the syscall
-   returns -1.
+   If :data:`WNOHANG` is specified and there are no matching children in the
+   requested state, ``(0, 0)`` is returned.
+
+   If there are no matching children that could be waited for,
+   :exc:`ChildProcessError` is raised.
```
Comment (aeros) [inline Doc/library/os.rst]:
```suggestion
   If :data:`WNOHANG` is specified and there are no matching children in the
   requested state, ``(0, 0)`` is returned. If there are no matching children that could be waited for,
   :exc:`ChildProcessError` is raised.
```
This was done for the previous section where `ChildProcessError` was mentioned. IMO, it reads a bit better as a single paragraph.

File: Doc/library/os.rst [hunk 2]
```diff
@@ -4041,83 +4041,44 @@ written in Python, such as a mail server's external command delivery program.
...
+.. function:: waitid(idtype, id, options)
 
-   .. versionadded:: 3.3
+   Wait for the completion of a child process.  *idtype* can be :data:`P_PID`,
+   :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD` on Linux.  The
+   interpretation of *id* depends on it, see their individual description.
 
+   *options* is an OR combination of flags.  At least one of :data:`WEXITED`,
+   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional
+   flags are :data:`WNOHANG` and :data:`WNOWAIT`.
```
Comment (aeros) [inline Doc/library/os.rst]: I think "At least one or more" is slightly better used here than just "At least one", to very explicitly convey that they can all be used in combination with one another. "one or more" was used in the previous documentation, as well in the man page. While both would be technically correct, IMO "one or more" is a bit more clear.

```suggestion
   *options* is an OR combination of flags.  At least one or more of :data:`WEXITED`,
   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional
   flags are :data:`WNOHANG` and :data:`WNOWAIT`.
```

File: Doc/library/os.rst [hunk 3]
```diff
@@ -4041,83 +4041,44 @@ written in Python, such as a mail server's external command delivery program.
...
-   Flags that can be used in *options* in :func:`waitid` that specify what
-   child signal to wait for.
-
-   .. availability:: Unix.
+.. function:: waitid(idtype, id, options)
 
-   .. versionadded:: 3.3
+   Wait for the completion of a child process.  *idtype* can be :data:`P_PID`,
+   :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD` on Linux.  The
+   interpretation of *id* depends on it, see their individual description.
```
Comment (aeros) [inline Doc/library/os.rst]: From a formatting perspective, I think it looks clearer to separate the brief description from the parameters.

(Also includes a minor grammar fix)

```suggestion
   Wait for the completion of a child process.  
   
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD`
   on Linux.  The interpretation of *id* depends on it, see their individual descriptions.
```

File: Doc/library/os.rst [hunk 4]
```diff
@@ -4173,17 +4137,121 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   These are the possible values for *idtype* in :func:`waitid`. They affect
+   how *id* is interpreted:
+
+   * With :data:`P_PID`, wait for the child with the PID given as *id*.
+   * With :data:`P_PGID`, wait for any child whose progress group ID is *id*.
+   * With :data:`P_ALL`, wait for any child; *id* is ignored.
+   * With :data:`P_PIDFD`, wait for the child identified by the file descriptor
+     *id* (a process file descriptor created with :func:`pidfd_open`). This is
+     Linux-specific.
```
Comment (aeros) [inline Doc/library/os.rst]: 1) Remove the "With" before each, and replace with just the bullet point and hyphen. This makes it more succinct.

2) Slightly reword `P_PID`. IMO, "wait for the child with the PID given as *id*." would read better as "wait for the child whose PID is *id*".

```suggestion
   * :data:`P_PID` - wait for the child whose PID is *id*.
   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
   * :data:`P_ALL` - wait for any child; *id* is ignored.
   * :data:`P_PIDFD` - wait for the child identified by the file descriptor
     *id* (a process file descriptor created with :func:`pidfd_open`).
```

File: Doc/library/os.rst [hunk 5]
```diff
@@ -4173,17 +4137,121 @@ written in Python, such as a mail server's external command delivery program.
...
+      The ``P_PIDFD`` constant.
+
+
+.. data:: WCONTINUED
+
+   This option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
+   :func:`waitid` causes child processes to also be reported if they have been
+   continued from a job control stop since their status was last reported.
+
+   .. availability:: some Unix systems.
```
Comment (aeros) [inline Doc/library/os.rst]: On Linux, `WCONTINUED` is specifically available on kernel 2.6.10+. However, I'm not certain that it's worth mentioning explicitly since 2.6.x has been EoL for a while now.

File: Doc/library/os.rst [hunk 6]
```diff
@@ -4041,83 +4041,44 @@ written in Python, such as a mail server's external command delivery program.
...
 
-.. data:: CLD_EXITED
-          CLD_KILLED
-          CLD_DUMPED
-          CLD_TRAPPED
-          CLD_STOPPED
-          CLD_CONTINUED
+   The return value is an object representing the data contained in the
+   :c:type:`siginfo_t` structure with the following attributes: :attr:`si_pid`,
+   :attr:`si_uid`, :attr:`si_signo`, :attr:`si_status`, :attr:`si_code`.
```
Comment (aeros) [inline Doc/library/os.rst]: It might be worth expanding here to explain each attribute, or better yet: a specific section for them similar to the `P_*` options. Upon building the documentation, it appears that these sphinx members don't currently link anywhere:

image

`si_pid`, `si_uuid`, and `si_signo` are reasonably intuitive, but it would be especially nice to have an explanation for `si_status` (`si_code` could probably just link to the `CLD_*` section). Until consulting the man page, I was personally unaware that it could contain the specific signal that resulted in the child process terminating, stopping, or continuing (my guess was that it only could return the exit status).

File: Doc/library/os.rst [hunk 7]
```diff
@@ -4041,83 +4041,44 @@ written in Python, such as a mail server's external command delivery program.
    number is zero); the high bit of the low byte is set if a core file was
    produced.
 
+   If there are no children that could be waited for, :exc:`ChildProcessError`
+   is raised.
```
Comment (aeros) [inline Doc/library/os.rst:4495]: I'm surprised `ChildProcessError` wasn't already mentioned, but it's definitely good to see it documented now.

File: Doc/library/os.rst [hunk 8]
```diff
@@ -4173,17 +4137,121 @@ written in Python, such as a mail server's external command delivery program.
...
+
+.. data:: CLD_EXITED
+          CLD_KILLED
+          CLD_DUMPED
+          CLD_TRAPPED
+          CLD_STOPPED
+          CLD_CONTINUED
+
+   These are the possible values for :attr:`si_code` in the result returned by
+   :func:`waitid`.
```
Comment (aeros) [inline Doc/library/os.rst]: Would it be worth including some brief details on each `si_code` value? 

I don't think it would be quite as useful as the `P_*` options, but some are not entirely intuitive. Particularly for `CLD_EXITED`, I would not guess that it's very specifically denoting that the child process called `_exit(2)` without consulting documentation. Others such as `CLD_KILLED`, `CLD_STOPPED`, and `CLD_CONTINUED` are significantly more obvious based on the name, but I think it would still be useful to have a short description with sphinx links to the `signal` module.

**Review by aeros — COMMENTED** [review]
Thanks for the significant improvements to the `os.wait*()` docs, @birkenfeld!

I verified the new `waitid()` content against the following [man page](https://man7.org/linux/man-pages/man2/waitid.2.html), as well as locally verifying the cases where `ChildProcessError` is raised. Other than a few specific parts where I have suggestions, it mostly looks good to me (it would be a very clear improvement as-is, so certainly feel free to include some or none of my suggestions if you disagree with any).

File: Doc/library/os.rst (hunk 6 above)
Comment (birkenfeld) [reply Doc/library/os.rst]: Indeed, they are not documented further, I can add that.

File: Doc/library/os.rst (hunk 4 above)
Comment (birkenfeld) [reply Doc/library/os.rst]: Nice!

File: Doc/library/os.rst (hunk 8 above)
Comment (birkenfeld) [reply Doc/library/os.rst]: > I would not guess that it's very specifically denoting that the child process called _exit(2) without consulting documentation

I don't think that's as significant as you think, since `_exit` is also what the C runtime calls when the `main` function finishes normally. So it really is just "exited normally".

**Review by aeros — APPROVED** [review]

**Review by iritkatriel — DISMISSED** [review]
This has merge conflicts now.

**Comment by birkenfeld** [comment]
I have made the requested changes; please review again

**Comment by iritkatriel** [comment]
The doc build has errors.

File: Doc/library/os.rst [hunk 9]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   * :data:`P_PID` - wait for the child whose PID is *id*.
+   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
+   * :data:`P_ALL` - wait for any child; *id* is ignored.
+   * :data:`P_PIDFD` - wait for the child identified by the file descriptor
+     *id* (a process file descriptor created with :func:`pidfd_open`).
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+   .. note:: ``P_PIDFD` is only available on Linux >= 5.4.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   .. note:: :data:`!P_PIDFD` is only available on Linux >= 5.4.
```

Fix the build error and add the appropriate syntax; I have a review in-work for the rest of it.

File: Doc/library/os.rst [hunk 10]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
    :func:`waitstatus_to_exitcode` can be used to convert the exit status into an
    exit code.
 
    .. availability:: Unix, not Emscripten, not WASI.
 
    .. seealso::
 
-      :func:`waitpid` can be used to wait for the completion of a specific
-      child process and has more options.
+      The other ``wait*`` functions documented below can be used to wait for the
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]: To use consistent syntax, formatting and semantics as other functions, but not try to resolve the reference, you could do:

```suggestion
      The other :func:`!wait*` functions documented below can be used to wait for the
```

Or, if you wanted to make it more explict that the `*` reprisents a wildcard, you could do

```suggestion
      The other :samp:`wait{*}` functions documented below can be used to wait for the
```

File: Doc/library/os.rst [hunk 11]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
-   Flags that can be used in *options* in :func:`waitid` that specify what
-   child signal to wait for.
+.. function:: waitid(idtype, id, options, /)
 
-   .. availability:: Unix, not Emscripten, not WASI.
+   Wait for the completion of a child process.
 
-   .. versionadded:: 3.3
+   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or :data:`P_PIDFD`
+   on Linux.  The interpretation of *id* depends on it, see their individual descriptions.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   on Linux.  The interpretation of *id* depends on it; see their individual descriptions.
```

Fix grammar error (comma splice)

Comment (CAM-Gerlach) [inline Doc/library/os.rst]: This states that *idtype* can be one of these options on Linux, but what about other *nix (macOS, BSD) which this function says it supports? The `availability` on these constants below also says `Unix`, which isn't consistent with just "Linux" stated here, so presumably one, the other or both should be clarified?

File: Doc/library/os.rst [hunk 12]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
 
+   *options* is an OR combination of flags.  At least one or more of :data:`WEXITED`,
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]: "At least one or more" is redundant, since "at least N" and "N or more" are synonymous. So, this should be either

```suggestion
   *options* is an OR combination of flags.  At least one of :data:`WEXITED`,
```

or

```suggestion
   *options* is an OR combination of flags.  One or more of :data:`WEXITED`,
```

File: Doc/library/os.rst [hunk 13]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
+   :data:`WSTOPPED` or :data:`WCONTINUED` is required.  Additional optional
+   flags are :data:`WNOHANG` and :data:`WNOWAIT`.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   :data:`WSTOPPED` or :data:`WCONTINUED` is required;
   :data:`WNOHANG` and :data:`WNOWAIT` are additional optional flags.
```

IMO, this reduces excessive choppiness and improves flow.

File: Doc/library/os.rst [hunk 14]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
-          CLD_DUMPED
-          CLD_TRAPPED
-          CLD_STOPPED
-          CLD_CONTINUED
+   The return value is an object representing the data contained in the
+   :c:type:`siginfo_t` structure with the following attributes: :attr:`si_pid`
+   (process ID), :attr:`si_uid` (real user ID of the child), :attr:`si_signo`
+   (always :data:`~signal.SIGCHLD`), :attr:`si_status` (the exit status or
+   signal number, depending on :attr:`si_code`), and :attr:`si_code` (see
+   :data:`CLD_EXITED` for possible values).
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   The return value is an object representing the data contained in the
   :c:type:`!siginfo_t` structure with the following attributes:
   
   * :attr:`!si_pid` (process ID)
   * :attr:`!si_uid` (real user ID of the child)
   * :attr:`!si_signo` (always :data:`~signal.SIGCHLD`)
   * :attr:`!si_status` (the exit status or signal number, depending on :attr:`!si_code`)
   * :attr:`si_code` (see :data:`CLD_EXITED` for possible values)
```

* Structure this as a bulleted list, which is a lot easier to read and quickly consult, as appropriate for a reference doc
* Make it more clear/less syntactically ambiguous that "with" refers to the object, not the structure or the data contained in it.
* De-resolve these attribute reference and the `:c:type:` reference, since they aren't actually documented as such so resolution fails, and results in warnings

File: Doc/library/os.rst [hunk 15]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
 
-   These are the possible values for :attr:`si_code` in the result returned by
-   :func:`waitid`.
+   If :data:`WNOHANG` is specified and there are no matching children in the
+   requested state, ``None`` is returned.  If there are no matching children
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   requested state, ``None`` is returned. 
   Otherwise, if there are no matching children
```

Add "otherwise", to clarify the relationship between the two (assuming that is correct), which took some time to parse as a reader.

File: Doc/library/os.rst [hunk 16]
```diff
@@ -4585,8 +4549,9 @@ written in Python, such as a mail server's external command delivery program.
    ``-1``, status is requested for any process in the process group ``-pid`` (the
    absolute value of *pid*).
 
-   An :exc:`OSError` is raised with the value of errno when the syscall
-   returns -1.
+   If :data:`WNOHANG` is specified and there are no matching children in the
+   requested state, ``(0, 0)`` is returned.  If there are no matching children
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   requested state, ``(0, 0)`` is returned.
   Otherwise, if there are no matching children
```

Likewise, make the relationship between these two a little more explicit and easier for the reader to parse.

File: Doc/library/os.rst [hunk 17]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+          P_ALL
+          P_PIDFD
+
+   These are the possible values for *idtype* in :func:`waitid`. They affect
+   how *id* is interpreted:
+
+   * :data:`P_PID` - wait for the child whose PID is *id*.
+   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
+   * :data:`P_ALL` - wait for any child; *id* is ignored.
+   * :data:`P_PIDFD` - wait for the child identified by the file descriptor
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   * :data:`!P_PID` - wait for the child whose PID is *id*.
   * :data:`!P_PGID` - wait for any child whose progress group ID is *id*.
   * :data:`!P_ALL` - wait for any child; *id* is ignored.
   * :data:`!P_PIDFD` - wait for the child identified by the file descriptor
```

These just refer to this very same entry, so are redundant and can just have resolution disabled, per current agreed convention.

**Comment by CAM-Gerlach** [comment]
Just FYI, the force-push invalidated my mostly-complete review, so it will be some additional time while I recover and remake it.

**Comment by iritkatriel** [comment]
> Just FYI, the force-push invalidated my mostly-complete review, so it will be some additional time while I recover and remake it.

Yeah, it's better to just push with additional commits (not force push). This way the reviewer can see what changed since the last review. We squash the commits when we merge so it doesn't matter.

File: Doc/library/os.rst [hunk 18]
```diff
@@ -4491,83 +4491,47 @@ written in Python, such as a mail server's external command delivery program.
...
 
    .. availability:: Unix, not Emscripten, not WASI.
 
    .. versionadded:: 3.3
 
-   .. versionchanged:: 3.9
-      Added :data:`CLD_KILLED` and :data:`CLD_STOPPED` values.
-
 
 .. function:: waitpid(pid, options, /)
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst:4539]: I can't reach it with a comment/suggestion, but while the prose documentation here and above states that `waitpid` is supported on both *nix and Windows (just with somewhat different behavior), the `availability` states that it is only on Unix. It seems the latter should be updated to `Unix, Windows, not Emscripten, not WASI.`

File: Doc/library/os.rst [hunk 19]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   * :data:`P_PID` - wait for the child whose PID is *id*.
+   * :data:`P_PGID` - wait for any child whose progress group ID is *id*.
+   * :data:`P_ALL` - wait for any child; *id* is ignored.
+   * :data:`P_PIDFD` - wait for the child identified by the file descriptor
+     *id* (a process file descriptor created with :func:`pidfd_open`).
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+   .. note:: ``P_PIDFD`` is only available on Linux >= 5.4.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   .. note:: :data:`!P_PIDFD` is only available on Linux >= 5.4.
```

Be consistent with formatting/semantics as the usages above, and with general modern convention.

File: Doc/library/os.rst [hunk 20]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   .. versionadded:: 3.3
+   .. versionadded:: 3.9
+      The ``P_PIDFD`` constant.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
      The :data:`!P_PIDFD` constant.
```

Likewise

File: Doc/library/os.rst [hunk 21]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+
+.. data:: WCONTINUED
+
+   This option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```

Make it more explicit what argument this is passed to, and use precise terminology to make explicit this is a flag value.

File: Doc/library/os.rst (hunk 18 above)
Comment (CAM-Gerlach) [reply Doc/library/os.rst:4539]: Also, the semantics of *options* are not described here, despite `wait3` and `wait4` explicitly referring to this function for a description of that argument. Presumably, it should be similar in form to the one in the previous `waitid` function, listing the flags that are required and optional (`WCONTINUED`, `WUNTRACED`, and `WNOHANG` are the available options; I'm guessing maybe at least one of the first tow is required, and the last one is optional like it is with `waitid`?)

File: Doc/library/os.rst [hunk 22]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   :func:`waitid` causes child processes to also be reported if they have been
+   continued from a job control stop since their status was last reported.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   :func:`waitid` causes child processes to be reported if they have been
   continued from a job control stop since they were last reported.
```

I found the syntax here quite hard to parse and understand, and simplified it a bit. Maybe it can be simplified further?

File: Doc/library/os.rst [hunk 23]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+
+.. data:: WEXITED
+
+   This option for :func:`waitid` causes child processes that have terminated to
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag for :func:`waitid` causes child processes that have terminated to
```

Clarify as above

File: Doc/library/os.rst [hunk 24]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   be reported.
+
+   The other ``wait*`` function always report children that have terminated,
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   The other ``wait*`` functions always report children that have terminated,
```

Fix grammar. You might also want to consider `` :func:`!wait*` `` or `` :samp:`wait{*}` `` for the syntax, to be more consistent with the formatting/semantics of individual function names and/or emphasize the variable portion more clearly.

File: Doc/library/os.rst [hunk 25]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   so this option is not available for them.
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+   .. versionadded:: 3.3
+
+
+.. data:: WSTOPPED
+
+   This option for :func:`waitid` causes child processes that have been stopped
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag for :func:`waitid` causes child processes that have been stopped
```

Clarify as above

File: Doc/library/os.rst [hunk 26]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   This option is not available for the other ``wait*`` functions.
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+   .. versionadded:: 3.3
+
+
+.. data:: WUNTRACED
+
+   This option for :func:`waitpid`, :func:`wait3`, and :func:`wait4` causes
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag for :func:`waitpid`, :func:`wait3`, and :func:`wait4` causes
```

Clarity as above

File: Doc/library/os.rst [hunk 27]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   current state has not been reported since they were stopped.
+
+   This option is not available for :func:`waitid`.
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+
+.. data:: WNOHANG
+
+   The option for :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag causes :func:`waitpid`, :func:`wait3`, :func:`wait4`, and
```

Clarify as above, and fix grammar

File: Doc/library/os.rst [hunk 28]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   :func:`waitid` to return immediately if no child process status is available
+   immediately.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]: To avoid the jarring repetition of "immediately", a few options:

```suggestion
   :func:`waitid` to return without waiting if no child process status is available
   immediately.
```

```suggestion
   :func:`waitid` to return right away if no child process status is available
   immediately.
```

```suggestion
   :func:`waitid` to return immediately away if no child process status is available
   right away.
```

```suggestion
   :func:`waitid` to return immediately if no child process status is available
   at the time the functions are called.
```

Or, if none of these work, you could at least do

```suggestion
   :func:`waitid` to return immediately if no child process status is
   immediately available.
```

File: Doc/library/os.rst [hunk 29]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+   .. availability:: Unix, not Emscripten, not WASI.
+
+
+.. data:: WNOWAIT
+
+   The option for :func:`waitid` to leave the child in a waitable state, so that
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   This *options* flag causes :func:`waitid` to leave the child in a waitable state, so that
```

As above, fix grammar and clarify

File: Doc/library/os.rst [hunk 30]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+   a later wait call can be used again to retrieve the child status information.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]: Fix grammar, and I was unclear on the intended meaning here with more explicit syntax—do you mean a later `os.waitid` call?

```suggestion
   a later :func:`!waitid` call can be used to retrieve the child status information again.
```

...a later call to any `wait*` function?

```suggestion
   a later :func:`!wait*` call can be used to retrieve the child status information again.
```

...or something else, e.g. a later call to the underlying OS API (which could be stated more clearly)?

File: Doc/library/os.rst [hunk 31]
```diff
@@ -4625,17 +4590,122 @@ written in Python, such as a mail server's external command delivery program.
...
+
+
+.. data:: CLD_EXITED
+          CLD_KILLED
+          CLD_DUMPED
+          CLD_TRAPPED
+          CLD_STOPPED
+          CLD_CONTINUED
+
+   These are the possible values for :attr:`si_code` in the result returned by
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   These are the possible values for :attr:`!si_code` in the result returned by
```

Like above, this isn't an actual documented attribute (and not module-level data), so we shouldn't try to resolve it, fail and issue a warning.

File: Misc/NEWS.d/next/Documentation/2020-09-22-12-32-16.bpo-41825.npcaCb.rst [hunk 32]
```diff
@@ -0,0 +1,3 @@
+Restructured the documentation for the ``os.wait*()`` family of functions,
```
Comment (CAM-Gerlach) [inline Misc/NEWS.d/next/Documentation/2020-09-22-12-32-16.bpo-41825.npcaCb.rst]:
```suggestion
Restructured the documentation for the :func:`os.wait* <os.wait>` family of functions,
```

Link the name to the baseline function in the category, so readers can jump to the appropriate docs more easily.

**Review by CAM-Gerlach — COMMENTED** [review]
Overall, this seems like a quite helpful change. I did have a some questions and comments to clarify points that I as a non-expert reader found ambiguous, confusing, unclear or potentially contradictory, and a number of suggestions with textual and reST syntax tweaks to fix typos/grammar issues, fix/improve the reST syntax, clarify the text and be consistent with current docs conventions. Thanks!

FYI, you might already be aware, but you can easily apply a bunch of suggestions at once by going to the `Files changed` tab, clicking `Add to batch` on each suggestion you want and then clicking `Commit`. You can also reply with your own modified suggestions, add those and commit them instead at the same time.

File: Doc/library/os.rst (hunk 11 above)
Comment (birkenfeld) [reply Doc/library/os.rst]: I think I meant `(a, b, c) or (d on Linux)`.
```suggestion
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or, on Linux, :data:`P_PIDFD`.
   The interpretation of *id* depends on it; see their individual descriptions.
```

File: Doc/library/os.rst (hunk 18 above)
Comment (birkenfeld) [reply Doc/library/os.rst:4539]: Yeah, will update the availability. `options` is described at least in that it should be `0` normally, which means no flags are required.

**Comment by birkenfeld** [comment]
Thanks for the review! All your suggestions/comments should be addressed.

File: Doc/library/os.rst [hunk 33]
```diff
@@ -4491,83 +4491,50 @@ written in Python, such as a mail server's external command delivery program.
...
-          CLD_KILLED
-          CLD_DUMPED
-          CLD_TRAPPED
-          CLD_STOPPED
-          CLD_CONTINUED
+   * :attr:`!si_pid` (process ID)
+   * :attr:`!si_uid` (real user ID of the child)
+   * :attr:`!si_signo` (always :data:`~signal.SIGCHLD`)
+   * :attr:`!si_status` (the exit status or signal number, depending on :attr:`!si_code`)
+   * :attr:`si_code` (see :data:`CLD_EXITED` for possible values)
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   * :attr:`!si_code` (see :data:`CLD_EXITED` for possible values)
```

Looks like this got missed somehow

File: Doc/library/os.rst [hunk 34]
```diff
@@ -4491,83 +4491,50 @@ written in Python, such as a mail server's external command delivery program.
...
-          WSTOPPED
-          WNOWAIT
+.. function:: waitid(idtype, id, options, /)
 
-   Flags that can be used in *options* in :func:`waitid` that specify what
-   child signal to wait for.
+   Wait for the completion of a child process.
 
-   .. availability:: Unix, not Emscripten, not WASI.
+   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or, on Linux, :data:`P_PIDFD`.
```
Comment (CAM-Gerlach) [inline Doc/library/os.rst]:
```suggestion
   *idtype* can be :data:`P_PID`, :data:`P_PGID`, :data:`P_ALL`, or (on Linux) :data:`P_PIDFD`.
```

Minor tweak to make the syntax a little more clear and avoid overloading the comma here.

**Review by CAM-Gerlach — APPROVED** [review]
Rechecked the source and the rendered output, and it LGTM now from my end, aside from a couple tiny nits. Thanks again!

**Comment by birkenfeld** [comment]
Sorry for missing these, added.

**Comment by CAM-Gerlach** [comment]
> Sorry for missing these, added.

No need to be sorry—it was me who missed one, if not both of them :)

**Review by CAM-Gerlach — APPROVED** [review]
LGTM from my end; thanks @birkenfeld !

**Comment by iritkatriel** [comment]
Thank you @birkenfeld, @aeros and @CAM-Gerlach.
